chaincode/marbles_transfer/marbles_transfer_queries.go

```diff
@@ -106,16 +114,78 @@ func queryMarblePrice(ctx contractapi.TransactionContextInterface, marbleID stri
 	return string(marblePriceJSON), nil
 }
 
-// TODO add a query to get all of an organization's proposed sales
-// Use GetPrivateDataByPartialCompositeKey to find all keys starting with typeMarbleForSale
-// hint: see sample at https://github.com/hyperledger/fabric-samples/blob/master/chaincode/marbles02/go/marbles_chaincode.go#L458
+// QueryAssetSaleAgreements returns all of an organization's proposed sales
+func (s *SmartContract) QueryAssetSaleAgreements(ctx contractapi.TransactionContextInterface) ([]Agreement, error) {
+	return queryAgreementsByType(ctx, typeMarbleForSale)
+}
+
+// QueryAssetBuyAgreements returns all of an organization's proposed buys
+func (s *SmartContract) QueryAssetBuyAgreements(ctx contractapi.TransactionContextInterface) ([]Agreement, error) {
+	return queryAgreementsByType(ctx, typeMarbleBid)
+}
+
+func queryAgreementsByType(ctx contractapi.TransactionContextInterface, agreeType string) ([]Agreement, error) {
+	collection, err := getClientImplicitCollectionName(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// Query for any object type starting with `agreeType`
+	agreementsIterator, err := ctx.GetStub().GetPrivateDataByPartialCompositeKey(collection, agreeType, []string{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to read from private data collection: %s", err.Error())
+	}
+	defer agreementsIterator.Close()
+
+	agreements := []Agreement{}
+
+	for agreementsIterator.HasNext() {
+		resp, err := agreementsIterator.Next()
+		if err != nil {
+			return nil, err
+		}
+
+		newAgree := new(Agreement)
+		err = json.Unmarshal(resp.Value, newAgree)
+		if err != nil {
+			return nil, err
+		}
+
+		agreements = append(agreements, *newAgree)
+	}
 
-// TODO add a query to get all of an organization's proposed buys
-// Use GetPrivateDataByPartialCompositeKey to find all keys starting with typeMarbleBid
+	return agreements, nil
```

Comment:
Just checking, will a valid json array get returned here?  I mean, do square brackets get added around the returned JSON agreements? If so, that's good.

Reply:
Yes

```
peer chaincode query -o localhost:7050 --ordererTLSHostnameOverride orderer.example.com --tls --cafile ${PWD}/organizations/ordererOrganizations/example.com/orderers/orderer.example.com/msp/tlscacerts/tlsca.example.com-cert.pem -C mychannel -n marbles_transfer -c '{"function":"QueryAssetBuyAgreements","Args":["marble1"]}'
[{"marble_id":"marble1","price":100}]
```